src/dft/backends/cufft/commit.cpp

```diff
@@ -147,126 +149,107 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+            if ((stride_vecs.vec_a[1] % stride_vecs.vec_a[2] != 0) ||
+                (stride_vecs.vec_b[1] % stride_vecs.vec_b[2] != 0)) {
                 throw mkl::unimplemented(
                     "dft/backends/cufft", __FUNCTION__,
-                    "cufft requires an input stride to be divisible by all smaller input strides!");
+                    "cufft requires a stride to be divisible by all smaller strides!");
             }
-            if (onembed[1] % onembed[2] != 0) {
-                throw mkl::unimplemented(
-                    "dft/backends/cufft", __FUNCTION__,
-                    "cufft requires an output stride to be divisible by all smaller output strides!");
-            }
-            inembed[1] /= inembed[2];
-            onembed[1] /= onembed[2];
+            stride_vecs.vec_a[1] /= stride_vecs.vec_a[2];
+            stride_vecs.vec_b[1] /= stride_vecs.vec_b[2];
         }
-        offsets[0] = config_values.input_strides[0];
-        offsets[1] = config_values.output_strides[0];
         const int batch = static_cast<int>(config_values.number_of_transforms);
         const int fwd_dist = static_cast<int>(config_values.fwd_dist);
         const int bwd_dist = static_cast<int>(config_values.bwd_dist);
 
         // When creating real-complex descriptions, the strides will always be wrong for one of the directions.
         // This is because the least significant dimension is symmetric.
-        // If the strides are invalid (too small to fit) then just don't bother creating the plan.
-        bool valid_forward = true;
-        bool valid_backward = true;
-        if (rank > 1) {
-            if (dom == dft::domain::REAL) {
-                valid_forward = (n_copy[rank - 1] <= inembed[rank - 1] &&
-                                 (n_copy[rank - 1] / 2 + 1) <= onembed[rank - 1]);
-                valid_backward = (n_copy[rank - 1] <= onembed[rank - 1] &&
-                                  (n_copy[rank - 1] / 2 + 1) <= inembed[rank - 1]);
+        // If the strides are invalid (too small to fit) then just don't bother creating the plan
+        auto check_stride_validity = [&](auto strides_fwd, auto strides_bwd) {
+            int inner_nfwd = n_copy[rank - 1]; // inner dimensions of DFT
+            // Complex data is stored conjugate even for real domains
+            int inner_nbwd = dom == dft::domain::REAL ? inner_nfwd / 2 + 1 : inner_nfwd;
+            int inner_sfwd = strides_fwd.back(); // inner strides of DFT
+            int inner_sbwd = strides_bwd.back();
+            bool valid = true;
+            for (int r = 1; r < rank; ++r) {
+                valid = valid && (inner_nfwd <= inner_sfwd) && (inner_nbwd <= inner_sbwd);
+                inner_nfwd *= n_copy[rank - r - 1];
+                inner_nbwd *= n_copy[rank - r - 1];
+                inner_sfwd *= strides_fwd[rank - r - 1];
+                inner_sbwd *= strides_bwd[rank - r - 1];
             }
```

Comment:
Does cuFFT support negative stride values? If yes, wouldn't these checks be invalid with negative strides?

Reply:
I *think* this is correct.

For CUFFT, `strides_fwd` and `strides_bwd` correspond to `inembed` and `onembed`, [where](https://docs.nvidia.com/cuda/cufft/index.html#advanced-data-layout)
>`n[i] ≤ inembed[i]`, `n[i] ≤ onembed[i]`

The strides can be negative, if you set `a_stride` and `b_stride` negative.